### ISSUE TITLE
HTTPProxy status takes a long time to update

### ISSUE THREAD

**Issue by erwbgy** [projectcontour/contour#3372]
**What steps did you take and what happened:**

In a cluster with more than 3000 HTTPProxy objects (and 2000+ clusters) the status of newly created HTTPProxy objects is regularly not set after 2 minutes.

**What did you expect to happen:**

HTTPProxy object status to be set within a few seconds or at least a reasonably finite amount of time.

**Anything else you would like to add:**

It looks like the status of HTTPProcy objects is updated [one-at-a-time](https://github.com/projectcontour/contour/blob/main/internal/dag/httpproxy_processor.go#L89).  Perhaps this could be done in parallel?

**Environment:**

- Contour version: 1.12.0
- Kubernetes version: (use `kubectl version`):  1.18.12
- Kubernetes installer & version: kubeadm 1.18.12
- Cloud provider or hardware configuration: on-prem
- OS (e.g. from `/etc/os-release`):  RHEL 8.2

**Comment by youngnick** [comment]
Thanks for logging this @erwbgy! 2 minutes is definitely longer than what I would expect as well.

There's no reason why the updates couldn't be parallelized, the way these updates work is that they are sent down a channel, with the object name and the update to perform, and you're right, the actual updates are all done by a single goroutine at the moment.

We could definitely speed this up by having multiple goroutines to pull the updates off the channel and do them, although we will need to be careful we don't hit rate limits on the apiserver (which are set pretty low by default!). Feels like we may need to make the number of updaters tunable so that you can trade off how fast the update queue drains vs how hard you hit your apiserver.

**Comment by erwbgy** [comment]
HTTPProxy objects are being created and deleted (though not always) continuously as users deploy their applications.  It is definitely related to the number of HTTPProxy objects but seems to be worse when there are lots of invalid objects such as when the Service has been deleted but the HTTPProxy object has not.

**Comment by youngnick** [comment]
I think this may have been related to #3411 and fixed by #3412. That's merged now, so if you are game to check with a `main` image tag, you could see. Otherwise, we are discussing releasing a patch for 1.13 with this and some other fixes, and we can check then.

**Comment by erwbgy** [comment]
Unfortunately I don't see the `level=debug msg="unsupported type received" context=StatusAddressUpdater` messages in the contour logs.

**Comment by youngnick** [comment]
Okay, thanks @erwbgy. Looks like we may need to add a worker pool to the status writer, and associated config and caveats. Let's talk about this one in the community meeting.

**Comment by skriss** [comment]
@erwbgy I believe we've made some major improvements here in more recent releases that you've seen based on your comments in another issue (can't find the link right now), so I'm going to close this out, but feel free to reopen or reach out again as needed!
